### PR TITLE
Penalize non agenda actions

### PR DESCRIPTION
**Pull request by pdasigi**
I see that penalizing non-agenda actions makes the model overfit less. The accuracy on a small validation set I've been experimenting with went down after we removed currying, but the coverage went up. After I made this change, we're back to having the same accuracy as we did with currying (63%.. not great) but with higher coverage, now close to 80%.
In addition to adding the option of penalizing non-agenda actions, I also moved the code for computing model scores and checklists into `take_step`, as recommended by @matt-gardner in the review of #926.

### CODE REVIEW

File: tests/models/encoder_decoders/nlvr_semantic_parser_test.py [hunk 1]
```diff
@@ -9,3 +14,35 @@ def setUp(self):
...
+                       {"left": ("", True, {}), "right": ("", False, {})},
+                       {"left": ("", True, {}), "right": ("", True, {})},
+                       {"left": ("", True, {}), "right": ("", False, {})}]
+        # Of the actions above, those at indices 0 and 2 are on the agenda, and there are padding
+        # indices at the end.
+        test_agenda = Variable(torch.Tensor([[0], [4], [-1], [-1]]))
+        # pylint: disable=protected-access
+        target_checklist, relevant_actions = self.model._get_checklist_target(test_agenda,
+                                                                              all_actions)
+        checklist_data = target_checklist.data.numpy()
```
Comment (matt-gardner) [inline tests/models/encoder_decoders/nlvr_semantic_parser_test.py]: Suggestion for brevity:

```python
numpy.testing.assert_almost_equal(target_checklist.data.numpy(), [[1], [0], [1]])
numpy.testing.assert_almost_equal(relevant_actions.data.numpy(), [[0], [2], [4]])
```

Just two lines, instead of six, and easier to read in a test - the relevant expected results are right next to each other.  Introducing new variables and having this large-looking code block makes me think the test is more complicated than it is.

File: tests/models/encoder_decoders/nlvr_semantic_parser_test.py [hunk 2]
```diff
@@ -9,3 +14,35 @@ def setUp(self):
...
+class NlvrDecoderStateTest(AllenNlpTestCase):
+    def test_get_checklist_balance(self):
+        # pylint: disable=no-self-use
+        fake_target = Variable(torch.Tensor([[1], [1], [1], [1], [0]]))
+        fake_checklist = Variable(torch.Tensor([[0], [2], [0], [1], [0]]))
+        # pylint: disable=protected-access
+        balance = NlvrDecoderState._get_checklist_balance(fake_target, fake_checklist)
+        balance_data = balance.data.numpy()
+        expected_balance_data = numpy.asarray([[1], [0], [1], [0], [0]])
+        numpy.testing.assert_array_equal(balance_data, expected_balance_data)
```
Comment (matt-gardner) [inline tests/models/encoder_decoders/nlvr_semantic_parser_test.py]: Same comment as above.  We also typically do `from numpy.testing import assert_almost_equal`, which makes the line shorter.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 3]
```diff
@@ -147,13 +151,19 @@ def forward(self,  # type: ignore
...
-        initial_checklist = nn_util.new_variable_with_size(agenda, agenda.size(), 0).float()
         # Each instance's agenda is of size (agenda_size, 1)
         agenda_list = [agenda[i] for i in range(batch_size)]
-        agenda_mask_list = [agenda_mask[i] for i in range(batch_size)]
-        # Each instance's checklist is of size (agenda_size, 1)
-        initial_checklist_list = [initial_checklist[i] for i in range(batch_size)]
+        checklist_targets = []
+        agenda_relevant_actions = []
+        initial_checklist_list = []
+        for batch_actions, batch_agenda in zip(actions, agenda_list):
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Done.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 4]
```diff
@@ -147,13 +151,19 @@ def forward(self,  # type: ignore
         labels_data = label.data.cpu()
         label_strings = [self.vocab.get_token_from_index(int(label_data), "denotations") for
                          label_data in labels_data]
-        # (batch_size, agenda_size, 1)
-        initial_checklist = nn_util.new_variable_with_size(agenda, agenda.size(), 0).float()
         # Each instance's agenda is of size (agenda_size, 1)
         agenda_list = [agenda[i] for i in range(batch_size)]
-        agenda_mask_list = [agenda_mask[i] for i in range(batch_size)]
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Removed it, thanks!

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 5]
```diff
@@ -218,6 +228,52 @@ def forward(self,  # type: ignore
...
         return outputs
 
+    def _get_checklist_target(self,
+                              agenda: torch.LongTensor,
+                              all_actions: List[ProductionRuleArray]) -> Tuple[torch.LongTensor,
+                                                                               torch.LongTensor]:
+        """
+        Takes an agenda and a list of all actions and returns a target checklist against which the
+        checklist at each state will be compared against to compute a loss, and the indices of
+        actions relevant for checklist loss computation (``relevant_actions``). If
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: I added that information here.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 6]
```diff
@@ -218,6 +228,52 @@ def forward(self,  # type: ignore
...
+        if self._penalize_non_agenda_actions:
+            terminal_indices = []
+            for index, action in enumerate(all_actions):
+                # Each action is a ProductionRuleArray, a dict with keys "left" and "right", and
+                # values are tuples where the second element shows whether element is a
+                # non_terminal.
+                if not action["right"][1]:
+                    terminal_indices.append(index)
+            terminal_indices_tensor = torch.Tensor([terminal_indices])  # (1, num_terminals)
+            # We want to return a checklist target and relevant actions that column vectors to make
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Fixed it, thanks! And changed `target_checklist` computation like you suggested.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 7]
```diff
@@ -454,11 +512,13 @@ class NlvrDecoderState(DecoderState['NlvrDecoderState']):
...
-        actions we want to see in the decoded output
-    agenda_mask : ``List[torch.LongTensor]``
-        List of masks corresponding to agendas that indicate unpadded action items
+    agenda_relevant_actions : ``List[torch.LongTensor]``
+        List of actions relevant for computing chacklist costs for instances, each of which is a
+        tensor containing the indices of the actions we want to see or not see in the decoded
+        output
+    checklist_target : ``List[torch.LongTensor]``
+        List of targets corresponding to agendas that indicate the states we want the checklists to
+        ideally be.
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: Done.

File: allennlp/models/encoder_decoders/nlvr_semantic_parser.py [hunk 8]
```diff
@@ -890,16 +899,56 @@ def take_step(self,  # type: ignore
...
-
+        considered_action_probs = nn_util.masked_softmax(action_logits, action_mask)
+        all_agenda_action_probs = state.get_agenda_action_probs()
+        # Mixing model scores and agenda selection probabilities to compute the probabilities of all
+        # actions for the next step and the corresponding new checklists.
+        # All action logprobs will keep track of logprob corresponding to each local action index
+        # for each instance.
+        all_action_logprobs: List[List[Tuple[int, torch.Tensor]]] = []
+        all_new_checklists: List[List[torch.LongTensor]] = []
+        for group_index, instance_info in enumerate(zip(state.batch_indices,
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/models/encoder_decoders/nlvr_semantic_parser.py]: I'll refactor this code really soon. I need to do that for running the two baseline models we talked about. I'll deal with this then. Adding a TODO.

File: tests/models/encoder_decoders/nlvr_semantic_parser_test.py [hunk 9]
```diff
@@ -9,3 +14,35 @@ def setUp(self):
...
         self.ensure_model_can_train_save_and_load(self.param_file)
+
+    def test_get_checklist_target(self):
+        # Creating a fake all_actions field where actions 0, 2 and 4 are terminal productions.
+        all_actions = [{"left": ("", True, {}), "right": ("", False, {})},
+                       {"left": ("", True, {}), "right": ("", True, {})},
+                       {"left": ("", True, {}), "right": ("", False, {})},
+                       {"left": ("", True, {}), "right": ("", True, {})},
+                       {"left": ("", True, {}), "right": ("", False, {})}]
+        # Of the actions above, those at indices 0 and 2 are on the agenda, and there are padding
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/models/encoder_decoders/nlvr_semantic_parser_test.py]: Yea, thanks!

File: tests/models/encoder_decoders/nlvr_semantic_parser_test.py [hunk 10]
```diff
@@ -9,3 +14,35 @@ def setUp(self):
...
+        expected_checklist_data = numpy.asarray([[1], [0], [1]])
+        numpy.testing.assert_array_equal(checklist_data, expected_checklist_data)
+        agenda_data = relevant_actions.data.numpy()
+        expected_agenda_data = numpy.asarray([[0], [2], [4]])
+        numpy.testing.assert_array_equal(agenda_data, expected_agenda_data)
+
+
+class NlvrDecoderStateTest(AllenNlpTestCase):
+    def test_get_checklist_balance(self):
+        # pylint: disable=no-self-use
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/models/encoder_decoders/nlvr_semantic_parser_test.py]: Done.

File: tests/models/encoder_decoders/nlvr_semantic_parser_test.py (hunk 1 above)
Comment (pdasigi) [reply tests/models/encoder_decoders/nlvr_semantic_parser_test.py]: Done, thanks!

File: tests/models/encoder_decoders/nlvr_semantic_parser_test.py (hunk 2 above)
Comment (pdasigi) [reply tests/models/encoder_decoders/nlvr_semantic_parser_test.py]: Done.
